src/backend/tests/unit/graph/graph/test_cycles.py

```diff
@@ -109,3 +114,99 @@ def test_cycle_in_graph_max_iterations():
+    cycle_vertices = find_cycle_vertices(graph._get_edges_as_list_of_tuples())
+    cycle_outputs_lists = [graph.vertex_map[vertex_id]._custom_component.outputs for vertex_id in cycle_vertices]
+    cycle_outputs = [output for outputs in cycle_outputs_lists for output in outputs]
+    for output in cycle_outputs:
+        assert output.cache is False
+
+    non_cycle_outputs_lists = [
+        vertex._custom_component.outputs for vertex in graph.vertices if vertex.id not in cycle_vertices
+    ]
+    non_cycle_outputs = [output for outputs in non_cycle_outputs_lists for output in outputs]
+    for output in non_cycle_outputs:
+        assert output.cache is True
+
+
+@pytest.mark.api_key_required
+def test_updated_graph_with_prompts():
+    # Chat input initialization
+    chat_input = ChatInput(_id="chat_input").set(input_value="bacon")
+
+    # First prompt: Guessing game with hints
+    prompt_component_1 = PromptComponent(_id="prompt_component_1").set(
+        template="Try to guess a word. I will give you hints if you get it wrong.\nHint: {hint}\nLast try: {last_try}\nAnswer:",
+    )
+
+    # First OpenAI LLM component (Processes the guessing prompt)
+    openai_component_1 = OpenAIModelComponent(_id="openai_1").set(
+        input_value=prompt_component_1.build_prompt, api_key=os.getenv("OPENAI_API_KEY")
+    )
+
+    # Conditional router based on agent response
+    router = ConditionalRouterComponent(_id="router").set(
+        input_text=openai_component_1.text_response,
+        match_text=chat_input.message_response,
+        operator="contains",
+        message=openai_component_1.text_response,
+    )
+
+    # Second prompt: After the last try, provide a new hint
+    prompt_component_2 = PromptComponent(_id="prompt_component_2")
+    prompt_component_2.set(
+        template="Given the following word and the following last try. Give the guesser a new hint.\nLast try: {last_try}\nWord: {word}\nHint:",
+        word=chat_input.message_response,
+        last_try=router.false_response,
+    )
+
+    # Second OpenAI component (handles the router's response)
+    openai_component_2 = OpenAIModelComponent(_id="openai_2")
+    openai_component_2.set(input_value=prompt_component_2.build_prompt, api_key=os.getenv("OPENAI_API_KEY"))
+
+    prompt_component_1.set(hint=openai_component_2.text_response, last_try=router.false_response)
+
+    # chat output for the final OpenAI response
+    chat_output_1 = ChatOutput(_id="chat_output_1")
+    chat_output_1.set(input_value=router.true_response)
+
+    # Build the graph without concatenate
+    graph = Graph(chat_input, chat_output_1)
+
+    # Assertions for graph cyclicity and correctness
+    assert graph.is_cyclic is True, "Graph should contain cycles."
+
+    # Run and validate the execution of the graph
+    results = []
+    max_iterations = 20
+    snapshots = [graph.get_snapshot()]
+
+    for result in graph.start(max_iterations=max_iterations, config={"output": {"cache": False}}):
```

Comment:
We probably want to make max_iterations required if the graph is cyclic?

Something like this in Graph.start()?
```
        if self.is_cyclic:
            if max_iterations is None:
                raise ValueError("max_iterations must be provided for cyclic graphs")
```

would we set other validations for max_iterations? > 0 and maybe a max value?

Reply:
Done. Thanks!